iceoryx_posh/include/iceoryx_posh/experimental/internal/popo/base_subscriber.inl

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IOX_EXPERIMENTAL_POSH_POPO_BASE_SUBSCRIBER_INL
+#define IOX_EXPERIMENTAL_POSH_POPO_BASE_SUBSCRIBER_INL
+
+namespace iox {
+namespace popo {
+
+template<typename T, typename port_t>
+BaseSubscriber<T, port_t>::BaseSubscriber(const capro::ServiceDescription& sd) : m_serviceDescription(sd)
+    /* : m_port(iox::runtime::PoshRuntime::getInstance().getMiddlewareReceiver(service, "")) */
+{}
+
+template<typename T, typename port_t>
+inline uid_t
+BaseSubscriber<T, port_t>::uid() const noexcept
+{
+    return m_uid;
+}
+
+template<typename T, typename port_t>
+inline cxx::expected<SubscriberError>
+BaseSubscriber<T, port_t>::subscribe(const uint64_t queueCapacity) noexcept
+{
+    m_subscriptionRequested = true;
+    uint32_t size = queueCapacity;
+    if (size > MAX_SUBSCRIBER_QUEUE_CAPACITY)
+    {
+        LogWarn() << "Cache size for subscribe too large " << size
+                  << ", limiting to MAX_RECEIVER_QUEUE_CAPACITY = " << MAX_SUBSCRIBER_QUEUE_CAPACITY;
+        size = MAX_SUBSCRIBER_QUEUE_CAPACITY;
+    }
+    m_port.subscribe(size);
+    return cxx::success<>();
+}
+
+template<typename T, typename port_t>
+inline SubscribeState
+BaseSubscriber<T, port_t>::getSubscriptionState() const noexcept
+{
+   return m_port.getSubscriptionState();
+}
+
+template<typename T, typename port_t>
+inline void
+BaseSubscriber<T, port_t>::unsubscribe() noexcept
+{
+    m_port.unsubscribe();
+    m_subscriptionRequested = false;
+}
+
+template<typename T, typename port_t>
+inline bool
+BaseSubscriber<T, port_t>::hasData() const noexcept
+{
+    return m_port.hasNewChunks();
+}
+
+template<typename T, typename port_t>
+inline cxx::optional<cxx::unique_ptr<T>>
+BaseSubscriber<T, port_t>::receive() noexcept
+{
+    auto result = m_port.tryGetChunk();
+    if(result.has_error())
+    {
+        /// @todo - what should we do when getting ChunkReceiveError ?
```

Comment:
I would say forwarding the error here. Then the user can realize that they already hold to many chunks